mpqp/core/instruction/measurement/expectation_value.py

```diff
@@ -7,61 +7,123 @@ class to define your observable, and a :class:`ExpectationMeasure` to perform
 
+if TYPE_CHECKING:
+    from cirq.circuits.circuit import Circuit as Cirq_Circuit
+
 from mpqp.core.instruction.gates.native_gates import SWAP
 from mpqp.core.instruction.measurement.measure import Measure
+from mpqp.core.instruction.measurement.pauli_string import PauliString
 from mpqp.core.languages import Language
 from mpqp.tools.errors import NumberQubitsError
+from mpqp.tools.generics import Matrix, one_lined_repr
 from mpqp.tools.maths import is_hermitian
-from mpqp.tools.generics import one_lined_repr
 
 
 @typechecked
 class Observable:
     """Class defining an observable, used for evaluating expectation values.
 
-    An observable can be defined by using a hermitian matrix, or using a combination of operators in a specific
-    basis (Kraus, Pauli, ...).
-
-    For the moment, on can only define the observable using a matrix.
+    An observable can be defined by using a Hermitian matrix, or using a combination of operators in a specific
+    basis Pauli.
 
     Example:
         >>> matrix = np.array([[1, 0], [0, -1]])
+        >>> pauli_string = 3 * I @ Z + 4 * X @ Y
         >>> obs = Observable(matrix)
+        >>> obs2 = Observable(pauli_string)
 
-    Args:
+    Attributes:
         matrix: Hermitian matrix representing the observable.
+        pauli_string: PauliString representing the observable.
+
+    Raises:
+        ValueError: If the input matrix is not Hermitian or does not have a square shape.
+        NumberQubitsError: If the number of qubits in the input observable does
+            not match the number of target qubits.
+
     """
 
-    def __init__(self, matrix: npt.NDArray[np.complex64] | list[list[Complex]]):
-        self.nb_qubits = int(np.log2(len(matrix)))
-        """Number of qubits of this observable."""
-        self.matrix = np.array(matrix)
-        """See parameter description."""
+    def __init__(self, observable: Matrix | PauliString):
+        """
+        Initialize an Observable object.
 
-        basis_states = 2**self.nb_qubits
-        if self.matrix.shape != (basis_states, basis_states):
-            raise ValueError(
-                f"The size of the matrix {self.matrix.shape} doesn't neatly fit on a"
-                " quantum register. It should be a square matrix of size a power"
-                " of two."
-            )
+        Args:
+            observable (Matrix | PauliString): The observable to be represented.
+                Can be specified as a matrix or a PauliString.
 
-        if not is_hermitian(self.matrix):
-            raise ValueError(
-                "The matrix in parameter is not hermitian (cannot define an observable)"
-            )
+        Raises:
+            ValueError: If the input matrix is not hermitian or does not have a
+                square shape.
+            NumberQubitsError: If the number of qubits in the input observable does
+                not match the number of target qubits.
+
+        """
+        self.observable = observable
```

Comment:
I'm not sure I agree with this, here is the way we see how we did it: the main (and could be the only) place data is stored is in this `observable` attribute. The private attributes are only used for caching.

In other words, having a single attribute allows us to have a single source of truth, which is quite nice :)

Reply:
although, you are right that is causes some unfriendly checks, in the `matrix` property for example. We'll see if it can be improved

Reply:
I agree with that, the problem is that when the user wants to retrieve the definition of the observable, matrix or PauliString, he will not know in advance which one it is, so he will have to perform some type check after that. If it is not the right definition, he will have no choice than calling the right property at the end.
Besides, I suppose that when the user will want to retrieve the observable definition, he knows in advance if he wants a matrix or a PauliString.
For thos two reasons I think it makes more sense to directly manipulate observable.matrix and observable.pauli_strings as properties, without having to pass from an intermediate attribute observable that I will have to process anyway

Reply:
hmmm, not really, `matrix` is always a well formed matrix and `pauli_string` is always a well formed PS, so no type checks needed. The only thing is `observable`, but we could even make this attribute private if it bothers you. Having a single source of truth is quite nice in my opinion

Reply:
> Besides, I suppose that when the user will want to retrieve the observable definition, he knows in advance if he wants a matrix or a PauliString.

I disagree with that though

Reply:
I meant that if the user retrieves the field ``observable`` of an instance of ``Observable`` he will have to check the type of the real definition stored there, because it is of type ``Union[PauliString, Matrix]``.

Maybe a quick example can illustrate my thoughts. Imagine I retrieve an ExpectationMeasure from a circuit and I want to compute the trace of the observable (supposing this feature is not implemented in ``Observable`` otherwise I can change the example), one would have to do this:
```python
obs = expectation_measure.observable
obs_definition = obs.observable
if isinstance(obs_definition, PauliString):
    obs_definition = obs.matrix
obs_trace = np.trace(obs_definition)
```

While by only using the properties directly, without using the field ``observable`` we can do:
```python
obs = expectation_measure.observable
obs_trace = np.trace(obs.matrix)
```

I don't see any case where the user would like to retrieve the definition of an ``Observable``, without having in mind if he wants to manipulate a ``PauliString`` or a ``Matrix``. If the user only wants to pass the observable from a place to another in the code, he can manipulate directly the object ``Observable``, without having to access the field ``observable``.

Reply:
I have a new argument in favor of keeping `observable`: 
Let's assume we get rid of it. The user inputs a PS, when retrieving the matrix, the object now has both a PS and a matrix. which one is the source of truth ? If the user wants to modify one of them, now the two wouldn't match. Hence the importance of a single source of truth.

Now, if leaving the observable is unfriendly in your opinion, we can move it to be a private attribute.

Reply:
No, you don't need to check for the observable. This would work:
```py
obs = expectation_measure.observable
obs_trace = np.trace(obs.matrix)
```
Because when we call .matrix, it will check and return the matrix even if it is a PauliString:
```py
    @property
    def matrix(self) -> Matrix:
        """
        Returns the matrix representation of the observable.

        Returns:
            np.ndarray: The matrix representation of the observable.
        """
        if isinstance(self.observable, PauliString):
            if self._matrix is None:
                self._matrix = self.observable.to_matrix()
            return self._matrix.astype(np.complex64)
        else:
            return self.observable
```
For the user, it doesn't change anything.

Reply:
After thinking about it, I agree that this "single source of truth" thing is not that obvious. Especially taking into account mutability. Indeed, since the `matrix` and `observable` properties return pointers to the private attributes. One could accidentally modify a private attribute, breaking the synchronization between the so called source of truth (`observable`) and the cached versions of it (`_matrix` and `_observable`). In this context the "best" solution is probably for the properties to return copies of the cached attributes. And considering the code required for this, `observable` would lose some of its relevance. Because of this, I agree with removing `observable` and having private attributes accessed and kept in sync using getters and setters.